File: python/sliding-window/permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        target = defaultdict(int)
        for i in s1:
            target[i] += 1

        current = defaultdict(int)
        left = 0
        for i in range(len(s2)):
            if s2[i] not in target:
                while left < i:
                    if s2[left] not in target:
                        left += 1
                        continue
                    else:
                        current[s2[left]] -= 1
                        left += 1
            else:
                current[s2[i]] += 1
                while current[s2[i]] > target[s2[i]]:
                    if s2[left] not in target:
                        left += 1
                        continue
                    else:
                        current[s2[left]] -= 1
                        left += 1
            if current == target:
                return True

        return False

File: c++/sliding-window/permutation-in-string.cc
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        std::unordered_map<char, int> count;
        for (char c : s1) {
            count[c] += 1;
        }

        int left = 0;
        for (int i = 0; i < s2.size(); ++i) {
            count[s2[i]] -= 1;

            if (count[s2[i]] == 0) {
                int zero = true;
                for (auto [ch, counter] : count) {
                    if (counter != 0) {
                        zero = false;
                        break;
                    }
                }

                if (zero) {
                    return true;
                }
            } else {
                while (count[s2[i]] < 0) {
                    count[s2[left]] += 1;
                    left += 1;
                }
            }
        }
        return false;
    }
};
